### ISSUE TITLE
[UX] Improve Class not found exception in autoloader

### ISSUE THREAD

**Issue by ofbeaton** [composer/composer#5337]
I propose that composer's autoloader improve the appearance of class not found exception messages by reminding the user that they could be missing a composer dependency. This would be in addition to, not instead of.

In a usual php project, a class not found error message means you forgot to declare that class, so the exception makes sense.

In a composer project, it often means that you forgot to include a dependency, which can happen when copy+pasting code from a tutorial or documentation. The exception would be a lot more useful if it reminded you of this other possible solution.

We all know that we can eventually figure it out, especially given the namespace is unlikely to be one for your application.

I don't know if composer could also know if it only installed the prod dependencies or not, but that would also be good information to add, since it is another common way for your to suddenly be missing a class in prod. (see https://github.com/symfony/symfony-standard/issues/975)

This improvement would change it from just being an error message to providing some hints on possible solutions. Could even possibly lower the noise in support channels for all projects using composer.

**Comment by kamilsk** [comment]
Maybe my case will be helpful for you:
1. create PHPUnit test extends https://github.com/kamilsk/Common/blob/3.x/src/Test/ClassAvailability.php (like this https://github.com/kamilsk/Common/blob/3.x/tests/ClassAvailabilityTest.php)
2. run `composer install -o`
3. run `phpunit`

**Comment by Seldaek** [comment]
I'm not strictly against this but it's pretty hard to achieve as the autoloading system is a stack of autoloaders and so we can't just throw in case out autoloader failed as there might be others following. That's really something that is best handled at the application level in an error handler or such. Unless someone has an idea how to implement this?

**Comment by staabm** [comment]
Maybe this type of information could be provided when we know that the composers autloader is the only one reigstered?

**Comment by Seldaek** [comment]
Yes maybe.. if that can be safely determined, but that sounds potentially wasteful tbh.. If it penalizes performances I don't think it's really worth it. If someone wants to investigate feel free.

**Comment by ofbeaton** [comment]
@Pierstoval mentioned in https://github.com/symfony/symfony-standard/issues/975
that if composer exposed if the dev dependencies were installed somehow this would allow downstream applications to maybe catch the error and make nicer errors for that common case.

**Comment by Pierstoval** [comment]
> In a usual php project, a class not found error message means you forgot to declare that class, so the exception makes sense.

I don't agree with this point, it can also come from the fact that the class is misspelled or that the class name/namespace differs from the file name/path (or vice-versa), or from the fact that you forgot to add the file to the VCS, or a wrong development strategy that induces a leak by allowing to require missing classes... I faced many of these issues in the past years because of "working too fast", and there is no "standard case".
There is no "usual php project" in this case, IMO.
 
Plus, what I proposed with exposing whether dev dependencies are installed is more a "logical" tool to avoid issues in prod environments when using `composer install --no-dev`, mostly to be used by third-party components.

I don't know how Composer could know that you asked for a class that is present in a dev-dependency, because it means that Composer would need to already know the classes/namespaces registered in this dependency, which is not possible because it is not installed.
What I proposed was to find a way to add "metadata" to Composer's autoloader (or anything accessible from any app using Composer's autoloader (with `require 'vendor/autoload.php';` for example, even a static class, a json file, etc.) that would help third-parties know which classes are registered.
 
A possible solution could be the use of the `psr-0` and `psr-4` information in every `composer.json` file to detect the namespace, maybe when resolving dependencies on an update or from the `composer.lock` file, but even then, I don't think it could work properly because of namespace mismatch or similarities, for example... :confused:
